scrapy/core/engine.py

```diff
@@ -257,19 +257,17 @@ def _schedule_request(self, request: Request, spider: Spider) -> None:
 
     def download(self, request: Request, spider: Optional[Spider] = None) -> Deferred:
         """Return a Deferred which fires with a Response as result, only downloader middlewares are applied"""
-        if spider is None:
-            spider = self.spider
-        else:
+        if spider is not None:
             warnings.warn(
                 "Passing a 'spider' argument to ExecutionEngine.download is deprecated",
                 category=ScrapyDeprecationWarning,
                 stacklevel=2,
             )
             if spider is not self.spider:
                 logger.warning("The spider '%s' does not match the open spider", spider.name)
-        if spider is None:
+        if self.spider is None:
             raise RuntimeError(f"No open spider to crawl: {request}")
-        return self._download(request, spider).addBoth(self._downloaded, request, spider)
+        return self._download(request, spider).addBoth(self._downloaded, request, spider)  # type: ignore
```

Comment:
I see…

It looks like we should change the typing of `Spider` in those 2 methods to `Optional[Spider]`, instead of ignoring the typing issue.

It also seems like `_download` should get the `if spider is None: spider = self.spider` logic.

Reply:
Yeah i think the same, done.